[Validator] Checking a BIC along with an IBAN

sylfabre opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #28166 
| License       | MIT
| Doc PR        | https://github.com/symfony/symfony-docs/pull/10349

A BIC comes usually with an IBAN so it's better to check that they are associated. This PR provides an `iban` option to `Symfony\Component\Validator\Constraints\Bic` to check the BIC against an IBAN.

It also provides an `ibanPropertyPath` to retrieves the IBAN using the property accessor like with comparison constraints.

nicolas-grekas commented on src/Symfony/Component/Validator/Constraints/Bic.php:
```diff
@@ -36,4 +39,23 @@ class Bic extends Constraint
+    public $ibanMessage = 'This Business Identifier Code (BIC) is not associated with IBAN {{ iban }}.';
+
+    public $iban;
+
```
I'd suggest to remove the blank lines between the added properties

nicolas-grekas commented on src/Symfony/Component/Validator/Constraints/Bic.php:
```diff
@@ -36,4 +39,23 @@ class Bic extends Constraint
+
+    public $ibanPropertyPath;
+
+    public function __construct($options = null)
```
missing "array" type hint?

nicolas-grekas commented on src/Symfony/Component/Validator/Constraints/BicValidator.php:
```diff
@@ -81,5 +92,43 @@ public function validate($value, Constraint $constraint)
+        } else {
+            $iban = $constraint->iban;
+        }
+        if ($iban) {
```
if (!$iban) => return early?

sylfabre replied:
@nicolas-grekas no because the logic of validators is to return `null` on the first violation.

If tomorrow we add some more checks after this new code then we don't want to return early in this check.

nicolas-grekas replied:
We shouldn't base current code on hypothetical future-based arguments. Here it would make the code clearer *now*.

nicolas-grekas commented on src/Symfony/Component/Validator/Constraints/BicValidator.php:
```diff
@@ -81,5 +92,43 @@ public function validate($value, Constraint $constraint)
+        if ($iban) {
+            $ibanCountryCode = substr($iban, 0, 2);
+            if (ctype_alpha($ibanCountryCode)) {
+                if (substr($canonicalize, 4, 2) !== $ibanCountryCode) {
```
the two "if"s should be combined in one using `&&`

nicolas-grekas commented on src/Symfony/Component/Validator/Constraints/BicValidator.php:
```diff
@@ -81,5 +92,43 @@ public function validate($value, Constraint $constraint)
+                        ->setCode(Bic::INVALID_IBAN_COUNTRY_CODE_ERROR)
+                        ->addViolation();
+
+                    return;
```
not needed

sylfabre replied:
@nicolas-grekas I can remove it but I've done like the other violations raised above

nicolas-grekas replied:
Please remove it yes, the code elsewhere is unrelated.

nicolas-grekas commented on src/Symfony/Component/Validator/Tests/Constraints/BicValidatorTest.php:
```diff
@@ -36,6 +57,113 @@ public function testEmptyStringIsValid()
+
+        if (method_exists($this, 'expectException')) {
+            $this->expectException(ConstraintDefinitionException::class);
+            $this->expectExceptionMessage(sprintf('Invalid property path "foo" provided to "%s" constraint', \get_class($constraint)));
```
I'm wondering: what's the full message like? How much repetition does it contain, if any?

sylfabre replied:
@nicolas-grekas I've copied-pasted the code from https://github.com/symfony/symfony/blob/master/src/Symfony/Component/Validator/Tests/Constraints/AbstractComparisonValidatorTestCase.php#L176

nicolas-grekas replied:
I'd still be interested in the answer :)

sylfabre replied:
@nicolas-grekas Invalid property path "foo" provided to "\Symfony\Component\Validator\Constraints\Bic" constraint

nicolas-grekas replied:
Sorry it looks like I wasn't explicit enough: on the line above that throws the exception, the message ends with [...] constaint`: %s`, where %s is replaced by some `$e->getMessage()`. My question is about the value of this %s message: what does it look like in this test case for example?

sylfabre replied:
@nicolas-grekas Neither the property "foo" nor one of the methods "getFoo()", "foo()", "isFoo()", "hasFoo()", "__get()" exist and have public access in class "Symfony\Component\Validator\Tests\Constraints\BicComparisonTest_Class".

sylfabre replied:
Neither the property "foo" nor one of the methods "getFoo()", "foo()", "isFoo()", "hasFoo()", "__get()" exist and have public access in class "Symfony\Component\Validator\Tests\Constraints\BicComparisonTest_Class"

nicolas-grekas reviewed (commented):
Thanks for working on this! Here are some random comments :)

nicolas-grekas commented:
Could you improve the PR and commit title please also? (I'd suggest to squash commits while doing so)

ro0NL commented on src/Symfony/Component/Validator/Constraints/BicValidator.php:
```diff
@@ -81,5 +92,38 @@ public function validate($value, Constraint $constraint)
+        } else {
+            $iban = $constraint->iban;
+        }
+        if (!$iban) {
```
shouldnt we prefer `nulll === $iban`  instead?

sylfabre replied:
@nicolas-grekas tell me whatever you prefer!

nicolas-grekas replied:
I'm fine with `!$iban` personally: the logic below also skips e.g. the empty string or other falsy values.

ro0NL replied:
right. different question then :)

shouldnt we expect a violation for arbitrary ibans?

nicolas-grekas replied:
you would get one by when adding a constraint on the iban field itself, isn't it enough?

nicolas-grekas commented:
Thanks, LGTM, except the test failure of course.

nicolas-grekas reviewed (approved):
thanks

nicolas-grekas commented on src/Symfony/Component/Validator/Constraints/BicValidator.php:
```diff
@@ -81,5 +92,38 @@ public function validate($value, Constraint $constraint)
+        } else {
+            $iban = $constraint->iban;
+        }
+        if (!isset($iban) || !$iban) {
```
I'd suggest to rewrite the previous this way
```php
        $iban = $constraint->iban;
        $path = $constraint->ibanPropertyPath;
        if ($path && null !== $object = $this->context->getObject()) {
            try {
                $iban = $this->getPropertyAccessor()->getValue($object, $path);
            } catch (NoSuchPropertyException $e) {
                throw new ConstraintDefinitionException(sprintf('Invalid property path "%s" provided to "%s" constraint: %s', $path, \get_class($constraint), $e->getMessage()), 0, $e);
            }
        }
        if (!$iban) { 
```

nicolas-grekas commented on src/Symfony/Component/Validator/Constraints/BicValidator.php:
```diff
@@ -91,5 +101,36 @@ public function validate($value, Constraint $constraint)
+    private function getPropertyAccessor()
+    {
+        if (null === $this->propertyAccessor) {
+            $this->propertyAccessor = PropertyAccess::createPropertyAccessor();
```
correct! should leverage the new LogicException introduced in #28513

sylfabre replied:
answered in previous comment

nicolas-grekas commented:
Thanks @ro0NL for the comments, good catch!
@sylfabre OK to you with the comments? They all make sense to me (please rebase also.)

Status: needs work

sylfabre commented on src/Symfony/Component/Validator/Constraints/Bic.php:
```diff
@@ -36,4 +39,20 @@ class Bic extends Constraint
     );
 
     public $message = 'This is not a valid Business Identifier Code (BIC).';
+    public $ibanMessage = 'This Business Identifier Code (BIC) is not associated with IBAN {{ iban }}.';
```
done

sylfabre commented:
@nicolas-grekas @ro0NL I've rebased and updated the PR as per your comments.

I've also updated accordingly the code I used as source for my feature

ro0NL commented on src/Symfony/Component/Validator/Constraints/BicValidator.php:
```diff
@@ -23,6 +26,13 @@
 {
+    private $propertyAccessor;
+
+    public function __construct(PropertyAccessor $propertyAccessor = null)
```
looking at `AbstractComparisonValidator` i dont think we use property access as a service, so this seems fine :+1:

ro0NL commented on src/Symfony/Component/Validator/Tests/Constraints/AbstractComparisonValidatorTestCase.php:
```diff
@@ -16,26 +16,6 @@
 use Symfony\Component\Validator\Exception\ConstraintDefinitionException;
 use Symfony\Component\Validator\Test\ConstraintValidatorTestCase;
 
-class ComparisonTest_Class
```
ok, i see why you added the new class on top as well in `BicValidatorTest` :)

i still think we prefer classes at the bottom, but technically this change applies to 3.4 and is not really related to your PR.

sylfabre replied:
@ro0NL should I remove my changes to these files then? and someone else makes another PR for this?

ro0NL replied:
i think we should avoid this conflict with lower (maintained) branches yes

xabbuh commented on src/Symfony/Component/Validator/Constraints/BicValidator.php:
```diff
@@ -93,5 +104,39 @@ public function validate($value, Constraint $constraint)
+    private function getPropertyAccessor(): PropertyAccessor
+    {
+        if (null === $this->propertyAccessor) {
+            if (!class_exists('Symfony\Component\PropertyAccess\PropertyAccess')) {
```
If I don't miss anything, this cannot happen as the constraint would have thrown the exception before.

ro0NL replied:
that would apply to all constraints where we move the check: #28644

sylfabre replied:
@nicolas-grekas should I then remove the exception here as it's already thrown in the constraint constructor?

sylfabre replied:
@nicolas-grekas I think this `getPropertyAccessor()` should be remove: we either use DI in the constructor to get the `PropertyAccess` component or use the static way to get it (like for Intl)

ro0NL replied:
the `BicValidator` has no `instanceof Bic` test (should be added though for consistency).

Technically this throw can still happen if called before `validate()` is called (unlikely, but possible from API pov)

I think it makes sense to keep the check; it's more easy to reason about, causes less confusion and makes no assumptions. And all constraints do it already, we can keep it even in 5.0. That's the shortest path towards consistency concerning accessing deps.

xabbuh commented on src/Symfony/Component/Validator/Constraints/Bic.php:
```diff
@@ -36,4 +39,20 @@ class Bic extends Constraint
+    public function __construct(array $options = null)
+    {
+        if (isset($options['iban']) && isset($options['ibanPropertyPath'])) {
+            throw new ConstraintDefinitionException(sprintf('The "%s" constraint requires only one of the "iban" or "ibanPropertyPath" options to be set, not both.', \get_class($this)));
```
I would reword this a bit:

> The "iban" and "ibanPropertyPath" of the Iban constraint cannot be used at the same time.

ro0NL commented on src/Symfony/Component/Validator/Constraints/Bic.php:
```diff
@@ -36,4 +39,20 @@ class Bic extends Constraint
+        }
+
+        if (isset($options['propertyPath']) && !class_exists(PropertyAccess::class)) {
+            throw new ConstraintDefinitionException(sprintf('The "%s" constraint requires the Symfony PropertyAccess component to use the "ibanPropertyPath" option.', \get_class($this)));
```
`Symfony\Component\Validator\Exception\LogicException` here

sylfabre committed:
[Validator] Checking a BIC along with an IBAN
Fix #28166

sylfabre commented:
> @sylfabre up to address the remaining comments?

I've rebased the branch onto master and there is one question for you

ro0NL commented on src/Symfony/Component/Validator/CHANGELOG.md:
```diff
@@ -13,6 +13,7 @@ CHANGELOG
  * deprecated using the `Bic`, `Country`, `Currency`, `Language` and `Locale` constraints without `symfony/intl`
  * deprecated using the `Email` constraint without `egulias/email-validator`
  * deprecated using the `Expression` constraint without `symfony/expression-language`
+ * validating a BIC based on a given IBAN
```
for discovery, id say:

- added options `iban` and `propertyPath` to `Bic` constraint

ro0NL commented on src/Symfony/Component/Validator/Constraints/Bic.php:
```diff
@@ -28,6 +30,7 @@ class Bic extends Constraint
     const INVALID_BANK_CODE_ERROR = '00559357-6170-4f29-aebd-d19330aa19cf';
     const INVALID_COUNTRY_CODE_ERROR = '1ce76f8d-3c1f-451c-9e62-fe9c3ed486ae';
     const INVALID_CASE_ERROR = '11884038-3312-4ae5-9d04-699f782130c7';
+    const INVALID_IBAN_COUNTRY_CODE_ERROR = '29a2c3bb-587b-4996-b6f5-53081364cea5';
```
`public const`, but feel free to ignore. This is a CS discussion also (we added public for new consts before though)

nicolas-grekas replied:
let's keep consistency

ro0NL commented on src/Symfony/Component/Validator/Constraints/Bic.php:
```diff
@@ -38,14 +41,25 @@ class Bic extends Constraint
+        }
+
+        if (isset($options['propertyPath']) && !class_exists(PropertyAccess::class)) {
+            throw new LogicException(sprintf('The "symfony/property-access" component is required to use the "%s" constraint with the "ibanPropertyPath" option.', \get_class($this)));
```
`with the "propertyPath" option`

sylfabre replied:
@ro0NL but the option really is `ibanPropertyPath`!

ro0NL reviewed (commented):
LGTM to in general :+1: we need to decide if we add the `$constraint instanceof Bic` and throw (BC break-ish) or trigger deprecation.

nicolas-grekas commented:
> we need to decide if we add the $constraint instanceof Bic and throw (BC break-ish) or trigger deprecation

easy to decide: no Bc breaks :)

OskarStark reviewed (approved):
Anything todo here?

ro0NL commented on src/Symfony/Component/Validator/Constraints/Bic.php:
```diff
@@ -38,14 +41,25 @@ class Bic extends Constraint
         }
 
+        if (isset($options['iban']) && isset($options['ibanPropertyPath'])) {
+            throw new ConstraintDefinitionException(sprintf('The "iban" and "ibanPropertyPath" of the Iban constraint cannot be used at the same time.', \get_class($this)));
```
`... options of the Bic constraint ...`

ro0NL reviewed (commented):
tests are red :)

ro0NL commented on src/Symfony/Component/Validator/Constraints/Bic.php:
```diff
@@ -38,14 +41,25 @@ class Bic extends Constraint
+    public $ibanPropertyPath;
 
-    public function __construct($options = null)
+    public function __construct(array $options = null)
```
not sure about this :)

sylfabre replied:
looks good for me as we expect an array.

Why aren't you sure about this?

xabbuh replied:
In this case, the change looks valid to me. Generally you can define a default option by overriding the `getDefaultOption()` method. But since that is not the case for the `Bic` constraint, it will already throw an exception when no array is passed.

ro0NL replied:
exactly, but now we'll never reach the exception at https://github.com/symfony/symfony/blob/491bbe741dc0762eeda59ceb7d441a520af8d878/src/Symfony/Component/Validator/Constraint.php#L136


so basically all constraints will throw this exception, except here it will cause a type error: https://3v4l.org/WjMuY

sylfabre commented:
@ro0NL tests are now green :)

xabbuh commented on src/Symfony/Component/Validator/CHANGELOG.md:
```diff
@@ -15,6 +15,7 @@ CHANGELOG
  * deprecated using the `Bic`, `Country`, `Currency`, `Language` and `Locale` constraints without `symfony/intl`
  * deprecated using the `Email` constraint without `egulias/email-validator`
  * deprecated using the `Expression` constraint without `symfony/expression-language`
+ * added options `iban` and `ibanPropertyPath` to Bic constraint
```
needs to be added to a new `4.3.0` section now as 4.2 is in feature freeze

xabbuh commented on src/Symfony/Component/Validator/Constraints/Bic.php:
```diff
@@ -38,14 +41,25 @@ class Bic extends Constraint
         }
 
+        if (isset($options['iban']) && isset($options['ibanPropertyPath'])) {
+            throw new ConstraintDefinitionException(sprintf('The "iban" and "ibanPropertyPath" options of the Iban constraint cannot be used at the same time.', \get_class($this)));
```
`self::class` instead of `get_class($this)`(same below)?

xabbuh commented on src/Symfony/Component/Validator/Constraints/BicValidator.php:
```diff
@@ -95,5 +106,39 @@ public function validate($value, Constraint $constraint)
+    private function getPropertyAccessor(): PropertyAccessor
+    {
+        if (null === $this->propertyAccessor) {
+            if (!class_exists('Symfony\Component\PropertyAccess\PropertyAccess')) {
```
we should use the `class` constant here

sylfabre replied:
@xabbuh like `PropertyAccess::class` ?
I haven't tried but I think it will throw an exception if the class does not exist, right?

xabbuh replied:
yes like that, and using the `class` constant works even when the class is not present: https://3v4l.org/knSo3

sylfabre replied:
@xabbuh thanks!

nicolas-grekas commented:
Thank you @sylfabre.
